apps/dsb-message-broker/src/utils/topic.schema.service.ts

```diff
@@ -1,53 +1,137 @@
         addFormats(this.ajv, { mode: 'fast', formats: ['date', 'time'], keywords: true });
+
+        const xsdDirectory = path.resolve(__dirname, `../../../xsd_files`);
+        if (!fs.existsSync(xsdDirectory)) {
+            fs.mkdirSync(xsdDirectory);
+        }
     }
 
-    public validate(
+    public async validateSchema(
+        fqcn: string,
+        topic: string,
+        schemaType: string,
+        schema: any
+    ): Promise<void> {
+        const { error } = await this.setValidator(fqcn, topic, schemaType, schema);
+        if (error) throw new Error(error);
+        return;
+    }
+    public async validate(
         fqcn: string,
         topic: string,
         payload: string
-    ): { isValid: boolean; error: any } {
+    ): Promise<{ isValid: boolean; error: any }> {
+        const validator = await this.getValidator(fqcn, topic);
+        if (!validator) return { isValid: true, error: null };
+        let isValid, error;
+
+        if (validator instanceof XMLValidator) {
+            const isWellformed = libxml.loadXmlFromString(payload);
+            libxml.loadSchemas([validator.pathToSchema]);
+            isValid = libxml.validateAgainstSchemas();
+
+            if (!isWellformed) {
+                error = JSON.stringify(libxml.wellformedErrors);
+            } else if (!isValid && libxml.validationSchemaErrors) {
+                error = JSON.stringify(Object.values(libxml.validationSchemaErrors)[0]);
+            }
+        } else {
+            try {
+                payload = JSON.parse(payload);
+            } catch (error) {
+                throw new BadRequestException(['Payload is not in JSON format', error.message]);
+            }
+
+            isValid = validator.validate(payload);
+            error = JSON.stringify(validator.validate.errors);
+        }
+
+        return { isValid, error };
+    }
+
+    private async setValidator(
+        fqcn: string,
+        topic: string,
+        schemaType: string,
+        schema: JSONSchemaType<any> | any
+    ): Promise<{ validator: any; error: any }> {
         let validators = this._validators.get(fqcn);
         if (!validators) validators = {};
+        let validator, error;
 
-        if (!validators[topic]) {
-            const channel = this.addressbook.getChannel(fqcn);
+        if (schemaType === 'XSD') {
+            const isWellformed = libxml.loadXmlFromString(schema);
 
-            const schema = channel?.topics?.find(
-                (_topic: any) => _topic.namespace === topic
-            )?.schema;
+            const pathToFile = await this.saveSchemaToFile(`${topic}.${fqcn}.xsd`, schema);
 
-            if (schema) {
-                const JSONSchema: JSONSchemaType<any> = JSON.parse(schema);
-                validators[topic] = this.ajv.compile(JSONSchema);
-                this._validators.set(fqcn, validators);
+            if (isWellformed) validator = new XMLValidator(schema, pathToFile);
```

Comment:
Remove the schema file after being used to not spam the dir with XSDs. Also as a filename, please use guid so you won't run into potential concurrency issues when handling multiple requests.

Reply:
Also, maybe proposal above won't work if there is some sort of cache based on those names.

Reply:
uuid is now used for file names.
i think deleting the xsd each time after message validated adds overload of creating xsd for each message validation. instead now the directory will be erased in mb instance creation.